// volvo-web-components/volvo-timeline/volvo-timeline-item.js
import { html } from "../../@polymer/polymer/polymer-element.js";
import "../../@polymer/polymer/lib/elements/dom-repeat.js";
import { VolvoElem } from "../volvo-elem/volvo-elem.js";
import "../volvo-icon/volvo-icon.js";
import "../volvo-theme/volvo-theme.js";
export class VolvoTimelineItem extends VolvoElem {
  static get is() {
    return 'volvo-timeline-item';
  }

  static get properties() {
    return {
      icon: {
        type: String,
        value: 'icon'
      },
      header: {
        type: String,
        value: 'header'
      },
      date: {
        type: Object,
        value: new Date()
      },
      info: {
        type: String,
        value: 'info'
      },
      eventType: {
        type: String,
        value: ''
      }
    };
  }

  static get observers() {
    return ['setEventTypeColor(eventType)'];
  }

  setEventTypeColor(eventType) {
    const circleElement = (this.shadowRoot || this).querySelector('.circle');
    const iconElement = (circleElement.shadowRoot || circleElement).querySelector('.icon');

    switch (eventType) {
      case 'DTC Event':
        circleElement.style.visibility = 'visible';
        circleElement.style.backgroundColor = 'var(--volvo-color-error-rgb-opaque)';
        iconElement.style.color = 'var(--volvo-color-error)';
        break;

      case 'Repair Event':
        circleElement.style.visibility = 'visible';
        circleElement.style.backgroundColor = 'var(--volvo-color-orange-rgb-opaque)';
        iconElement.style.color = 'var(--volvo-color-orange)';
        break;

      default:
        circleElement.style.visibility = 'hidden';
    }
  }

  static get version() {
    return "0.3.2";
  }

  static get template() {
    return html`<style>\n  :host {\n    display: block;\n  }\n  .container {\n    position: relative;\n    width: 252px;\n    height: 54px;\n    background-color: white;\n\n    border-width: 1px;\n    border-style: solid;\n    border-color: var(--volvo-color-fog);\n    border-radius: 3px;\n  }\n  .header {\n    position: absolute;\n    top: 15%;\n    left: 21%;\n    @apply --volvo-text-body-small;\n  }\n  .date {\n    position: absolute;\n    bottom: 16%;\n    left: 21%;\n    @apply --volvo-text-caption-small;\n  }\n  .info {\n    position: absolute;\n    bottom: 16%;\n    right: 4.5%;\n    @apply --volvo-text-caption-small;\n  }\n  .circle {\n    position: absolute;\n    top: 25%;\n    left: 4%;\n    width: 30px;\n    height: 30px;\n    border-radius: 50%;\n    background-color: var(--volvo-color-stop);\n\n    display: flex;\n    justify-content: center;\n    align-items: center;\n  }\n  .icon {\n    max-width: 100%;\n    max-height: 100%;\n    top: -1px;\n    bottom: 0;\n    left: 1px;\n    right: 0;\n    margin: 0;\n  }\n</style>\n<div class="container">\n  <div class="circle">\n    <volvo-icon class="icon" icon="[[icon]]"></volvo-icon>\n  </div>\n  <div class="header">[[header]]</div>\n  <volvo-time class="date" datetime="[[date]]" date-only></volvo-time>\n  <div class="info">[[info]]</div>\n</div>\n`;
  }

}
window.customElements.define(VolvoTimelineItem.is, VolvoTimelineItem); //# sourceMappingURL=volvo-timeline-item.js.map
